### DISCUSSION TITLE
Q&A from the April 2024 USWDS monthly call

### DISCUSSION THREAD

**Discussion by finekatie** [uswds/uswds#5922] [Past community calls]
> **Community update — September 17, 2026:** This is a historical monthly-call post. For current questions and participation, use [USWDS GitHub Discussions](https://github.com/uswds/uswds/discussions). Recordings remain available in the [monthly call archive](https://designsystem.digital.gov/about/monthly-calls/). References below to Slack, staff roles, or future calls reflect the community at the time.

At the [April 2024 monthly call](https://digital.gov/event/2024/04/18/uswds-monthly-call-april-2024/), USWDS developers talked about Web Components technology: what it is and what it can mean for the design system. We discussed the value of bringing Web Components into the design system and exploring how they can affect all users. 

**Note:** These monthly call Q&A recap posts are still a new feature here. We’re still experimenting with what works best for these in the interest of transparency, but also how well they help you. Let us know if you find them useful — or if you have a better idea. Also, if you don’t see your question about the April call answered, feel free to discuss further here. Keep in mind this covers what we said during call Q&A as well as questions we didn’t have time for, and they’ve been lightly edited, but things may have changed. In those cases, we indicate updates with a date (5-10-24, for example).

### Questions from the April 2024 monthly call

- [Could you talk a little more about the considerations for multiple languages when you are planning Web Components?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9388978)
- [Will there be a public repo soon? Do you want contributors?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9388993)
- [I've been developing Web Components (and integrating them in Drupal) for the last three years. How can I help?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9388999)
- [May we customize the alert headers, so that instead of "Error Status," "Info Status," or "Alert status" they state specific errors, information, and alert titles? Our engineers feel they need to use the boilerplate default headers.](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389011)
- [Could you please make a connection between Web Components and digital experience indicators?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389018) 
- [What do you envision for plain language influences?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389029)
- [Does the move to Web Components mean a move towards more flexible components? For instance, the side nav is basically a list component, but it can’t be used as a list component because it’s built to be side navigation. Will a more generic list component be built to fill the role of side nav _and_ the role of menu, and the role of a content feed etc.?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389034)
- [Do you plan to continue to use SASS for CSS processing, or is that up in the air, too?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389042)
- [Will this shift also include JSON design tokens? Have you looked into moving from hex number to OKLCH?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389050)
- ​​[What are the options you're looking at for Web Components?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389052)
- [Will Web Components live alongside existing USWDS components? Or will this be a departure? Will I need to convert all my components to Web Components? Will I need to do it all at once?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389062)
- [Will you be using Shadow DOM?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389071)
- [How does USWDS alert agency and department project teams about specific component updates?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389083)
- [What will updating look like in this Web Components future?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389090)
- [Will the accessibility checklists also work with the Web Components?](https://github.com/uswds/uswds/discussions/5922#discussioncomment-9389098)

**Comment by finekatie** [comment]
**Q: Could you talk a little more about the considerations for multiple languages when you are planning Web Components?**

A: We have a few goals for language support. Our Web Components have the following requirements:

- To support the existing Spanish variants we offer 
- To use the lang attribute to easily switch between existing language variants
- To allow teams to add their own language support

At this time, we don't know how implementation would work. We might allow teams to load their own translations into the component via something like JSON, or to use component slots to add that content.

**Comment by finekatie** [comment]
**Q: Will there be a public Web Components repo soon? Do you want contributors?**

A: Our goal is to deliver something in an alpha form ASAP. We’ll be publishing a separate Web Components repo for clarity. We absolutely think there’s a role for contributors to help build things out and work through these issues, and we hope that in about a month we'll publish this public repo and one component.

**Update (5-10-24):** We have a new working repo at [github.com/uswds/uswds-next](http://github.com/uswds/uswds-next)

**Reply by angeliquejw** [reply]
This is an exciting update; the `uswds-next` repo is currently private -- is that intentional?

**Reply by thisisdano** [reply]
It's public now!

**Reply by thisisdano** [reply]
Thanks for the prompt to flip the switch

**Comment by finekatie** [comment]
> **Community update — September 17, 2026:** The Slack guidance below is historical. For current questions and participation, use [USWDS GitHub Discussions](https://github.com/uswds/uswds/discussions). Past calls remain available in the [monthly call archive](https://designsystem.digital.gov/about/monthly-calls/).

**Q: I've been developing Web Components (and integrating them in Drupal) for the last three years. How can I help?**

A: Contributions are welcome. We’d love to learn more about how you already use Web Components in the real world. Among other things, we’re always interested in how to handle component variants, passing content & data into components, multilingual support, and theming. Share your experience in [GitHub discussions](https://github.com/uswds/uswds/discussions) or [Slack](https://designsystem.digital.gov/about/community/).

**Comment by finekatie** [comment]
**Q: Can we customize the alert headers, so that instead of "Error Status," "Info Status," or "Alert status" they state specific errors, information, and alert titles? Our engineers feel they need to use the boilerplate default headers.**

A: Yes, you can customize. The [alert usability guidance](https://designsystem.digital.gov/components/alert/#usability-guidance-alert) recommends providing necessary context in plain language that avoids jargon or unnecessary technical language.

**Comment by finekatie** [comment]
**Q: Could you please make a connection between Web Components and digital experience indicators?** 

A: We want a reliable way to understand when sites are using USWDS components, and that could potentially be used in these indicators. Right now, it’s tricky to scan a website and determine whether it’s using something like the USWDS banner since there are a lot of different ways a team could implement it. With the move to Web Components, we hope to make USWDS implementation more consistent, which could make it easier to determine the presence of specific components.

**Comment by finekatie** [comment]
**Q: What do you envision for plain language influences?**

A: The connection between Web Components and plain language is a bit indirect. The design system won’t have a lot of material on plain language. Our goal is to make it easy for teams to add plain language content to any component. Understanding your audience, alongside good writing and structure, is critical to an effective service. We want our components to support effective, plain language writing.

There are some components that contain specific, consistent language, like the [Banner](https://designsystem.digital.gov/components/banner/). Web Components technology will make it easier for teams to stay up-to-date with any content improvements we make to the banner, and this certainly includes improvements to plain language.

**Comment by finekatie** [comment]
**Q: Does the move to Web Components mean a move towards more flexible components? For instance, the side nav is basically a list component, but it can’t be used as a list component because it’s built to be side navigation. Will a more generic list component be built to fill the role of side nav and the role of menu, and the role of a content feed, etc.?**

A: Components are created to fill a specific role, and a content feed might merit its own component and role. We recommend keeping components’ code encapsulated and separate from each other. If a generic, easy-to-maintain list component could meet multiple needs, we’d consider it, but we're a little skeptical of building extremely mutable components that parse component attributes to render very different outputs. Our [GitHub component proposals discussion board](https://github.com/uswds/uswds/discussions/categories/proposals) is a great place to talk about this further.

**Comment by finekatie** [comment]
**Q: Do you plan to continue to use Sass for CSS processing, or is that up in the air, too?**

A: Yes, in the short term we will continue to use Sass to style our traditional HTML components. We may also use Sass to power the styling for Web Components, but this is still under discussion. Long-term, we’ll be looking at ways to use more native CSS to reduce our dependence on custom processors.

**Comment by finekatie** [comment]
**Q: Will this shift also include JSON design tokens? Have you looked into moving from hex number to OKLCH?**

A: Yes, JSON design tokens are an important part of the shift to Web Components. We want to support the existing Sass codebase as well as a Web Components implementation that may be better integrated with CSS custom properties (CSS variables). 

Right now we have no opinion on what color space or model we use for our color output. The important thing for us now is the color's relationship to a specific design token. If there are practical reasons to switch color spaces to something like OKLCH, we'd be interested in that discussion.

**Comment by finekatie** [comment]
**Q: What are the options you're looking at for Web Components?**

A: We looked at native methods and some frameworks very close to native. We’d like to stay flexible and also support teams using other frameworks, like Angular or React.

We're looking at Open Web Components library, Lit, and StencilJS. For now, we're spending time with StencilJS because it gives us the opportunity to collaborate with teams who are using Web Components and StencilJS, but this is still an active discussion.

**Reply by anselmbradford** [reply]
Hi, has this answer evolved? I notice in https://github.com/uswds/uswds-next that the alpha components are all Lit-based. Is this the direction things are going, or is it yet to be determined?

**Reply by heymatthenry** [reply]
We're moving forward with Lit for the time being. There's a [decision record](https://github.com/uswds/uswds-proposals/pull/7) in progress that explains some of the reasoning behind that, but the gist of it is that:

- Lit is close to the platform; and 
- the switching costs would be fairly low if we decided at some point in the future that we needed to choose something else.

**Comment by finekatie** [comment]
**Q: Will Web Components live alongside existing USWDS components? Or will this be a departure? Will I need to convert all my components to Web Components? Will I need to do it all at once?**

A: We see Web Components existing at the same time as the current Version 3 of USWDS. We won't be switching Web Components on and traditional USWDS off. We expect to develop Web Components in a separate repo and as a separate package. We don’t want to immediately wind down our existing infrastructure that so many teams depend on. 

We want teams to be able to move to Web Components incrementally, replacing parts over time. You shouldn’t have to choose all one or the other.

**Comment by finekatie** [comment]
**Q: Will you be using Shadow DOM?**

A: Yes, especially for components where we want to minimize required changes for teams like possibly usa-banner and potentially usa-identifier. These core components contain a lot of markup that may not be easy for teams to update and maintain. With this in mind, we’ll need to be careful about any accessibility implications. There's a balancing act here, and we take both maintainability and component resilience seriously.

**Comment by finekatie** [comment]
**Q: How does USWDS alert agency and department project teams about specific component updates?**

A: You’ll find several ways to keep current: 

- Our [What's new](https://designsystem.digital.gov/about/whats-new/) page
- Our [Releases page](https://github.com/uswds/uswds/releases) on GitHub
- GitHub [discussions board](https://github.com/uswds/uswds/discussions) for release discussions
- Each component has a record of changes made in a “Latest updates” section, like this [example on the accordion component page](https://designsystem.digital.gov/components/accordion/#latest-updates).

**Comment by finekatie** [comment]
**Q: What will updating look like in this Web Components future?**

A: In general, our goal is to make it simpler to update from version to version. Just as with the current distribution model, you'd be able to install or update to the latest version of the distribution package via npm. As long as our updates don't affect the component API, updates should be simple and straightforward. And one of the goals of our move to Web Components is to develop simple, reliable component APIs that don't change much. 

There could be other distribution methods as well. We are starting to investigate what it could mean to deliver the design system via something like a CDN. Simplicity, reliability, performance, predictability, and security are key priorities as we consider possibilities for updating. This is another place where we'd be open to more discussion.

**Comment by finekatie** [comment]
**Q: Will the accessibility checklists also work with the Web Components?**

A: Yes. Web Components will have the same accessibility baseline as our current components.
